### DISCUSSION TITLE
Should Saved Searches be put into a separate app?

### DISCUSSION THREAD

**Discussion by sifex** [SigmaHQ/pySigma#19] [General]
This line here where pySigma starts to output to Splunk `savedsearch.conf` configuration file might be a step to far with what pySigma should be responsible for.

https://github.com/SigmaHQ/pySigma/blob/a53094445053b6f3d0bf0d32eb689cf183ca8055/sigma/conversion/backends/splunk.py#L86

Imo, pySigma should remain as un-opinionated as possible, but more concretely, outputting to `savedsearch.conf` w/o a method of setting other parameters, including :

- `alert.expires`
- default cron schedule (when `detection.timeframe` is not specified) , 
- relation, quantity & countype, and 
- actions such as webhook, 

poses bloat if pySigma accepts responsibility to support outputting this way. Additionally you'd be having to track savedsearch schema changes across Splunk versions, as well as providing further _separate_ configuration file dependency.

My suggestion would be that it's better to stop at the SPL layer to prevent other SIEM advocates using this as precedent to support their integration / configuration, and instead encouraging the community to create a pySigma version of a Sigma to Splunk TA / Splunk App that can have an opinionated implementation, along with supporting Dashboards, independent Configuration Files for compiling, etc.

Let me know your thoughts

**Comment by thomaspatzke** [comment]
I agree and would even go further and ask the question if any backends should be implemented inside pySigma or as separate projects.

The current situation with regard to the backend implementation in the sigmatools package is unsatisfactory for various reasons:

- Implementation in the core project suggests an ownership and responsibility, which is often not the case. Most backend code is maintained by individual contributors or organizations while the core is mostly maintained by myself. Some backends appear to be abandoned and don't seem to be maintained anymore. Separating them in different projects would make the state (especially if a backend if PoC, stable or abandoned) and responsbility for different code bases more obvious.
- Currently, issues for rules, core and backends are opened in one project and I think it's quite inconvenient for a contributor who took over a particular part to monitor the issues for the whole thing.
- Dependencies only required by a particular backend are inherited by the whole toolchain.

Therefore my suggestion is to completely separate the pySigma core library and backends into separate projects/packages. This would solve the issues mentioned above, but would also add some new issues:

- Currently it's quite easy for an user to install a working version of the conversion tool. It would be desirable to keep this state in the future, too. E.g. sigma-cli could ask the user to install a backend package if the user intends to use it without being installed. What do you think, @sifex?
- Versioning gets more challenging because backends will rely on pySigma features which get implemented over time. This could cause unresolvable conflicts, e.g. if backends depend on incompatible versions of pySigma. In such situations it would get hard to put them under the same CLI. Stabilization of the core, strict semantic versioning and backwards compatibility gets very important.
- In the current development state of pySigma, development of backends often causes changes/additions to the core, which can be directly implemented there. I would like to keep this state until the core has stabilized for the above reasons.

Short, the roadmap would be:

1. Finalize development of the pySigma core together with some backends (I have a list of some with challenges that also have to be solved at the core level) in one code base.
2. After the core has stabilized, splitting off the backends into separate projects.

What do you think about this plan?

**Reply by sifex** [reply]
> - Implementation in the core project suggests an ownership and responsibility, which is often not the case. Most backend code is maintained by individual contributors or organizations while the core is mostly maintained by myself. Some backends appear to be abandoned and don't seem to be maintained anymore. Separating them in different projects would make the state (especially if a backend if PoC, stable or abandoned) and responsbility for different code bases more obvious.
> - Currently, issues for rules, core and backends are opened in one project and I think it's quite inconvenient for a contributor who took over a particular part to monitor the issues for the whole thing.

1000% 👍 . In my Sigma-to-Splunk app, I found a 500%-800% speed increase by eliminating all other backends (`yaml.safe_load` is an incredibly slow operation when it's over all backends every run)

> What do you think about this plan?

Yea I like this approach. My first thought fell to something like [flysystem](https://flysystem.thephpleague.com/v2/docs/) which is a storage ecosystem that has many third party 'adapters' which are maintained by the community. 

I think creating an example "templated" backend python project to show what a third party backends look like would be good (Incl. Unit Tests, creating standardised and user-configurable means of completing `logsource` and `field` mappings, limitations of the SIEM, more info provided by a maintainer who specifically uses the backend professionally). As you mentioned, it does raise the challenge of settling on major versions of the API that guarantees backwards compatibility. It would create limitations such as:

- Not add too many modifiers / features during a major version, as backend maintainers can't be expected to adopt them in a timely manor.
- Not to deprecate any existing modifiers, but instead provide warnings that modifiers will be deprecated in future versions. 

>  E.g. sigma-cli could ask the user to install a backend package if the user intends to use it without being installed.

Too easy, as long as you're happy for `sigma-cli` to contain links to each backends & their respective python package, and sigma-cli would call `poetry` or `pipenv` and install one depending on the interactive CLI / CLI params. I think a centralized list would be required and backend maintainers would submit PRs to add to the list, similar to how [flysystem](https://flysystem.thephpleague.com/v2/docs/) does it. Of course if they'd prefer to `pipenv install` backends personally, that _should_ be able to be supported as well, but I wouldn't know the exacts.

**Comment by thomaspatzke** [comment]
Regarding different output formats in a backend as implemented for Splunk, the intention here was to reduce the inflationary creation of different backends for mainly the same query language in different output formats. I think it's hard to put a concrete border after which opinionated output begins. In fact, there are often various possibilities to express something in simple queries and we already had discussions in the past about how a qeury should look alike.

**Reply by sifex** [reply]
Yea, that makes sense. In that case I think it's fair to say that the `splunk` backend maintainer would be in charge of the available outputs formats that it can provide. Not yet sure if `sigma-cli` would provide that interface or not under the `-of / --output-format` flag.
